### PR TITLE
[GlobalISel] Refactor Combiner MatchData & Apply C++ Code Handling

### PR DESCRIPTION
**Pull request by Pierre-vh**
Combiners that use C++ code in their "apply" pattern only use that. They never mix it with MIR patterns as that has little added value.

This patch restricts C++ apply code so that if C++ is used, we cannot use MIR patterns or builtins with it. Adding this restriction allows us to merge calls to match and apply C++ code together, which in turns makes it so we can just have MatchData variables on the stack.

So before, we would have
```
  GIM_CheckCxxInsnPredicate // match
  GIM_CheckCxxInsnPredicate // apply
  GIR_Done
```
Alongside a massive C++ struct holding the MatchData of all rules possible (which was a big space/perf issue).

Now we just have
```
GIR_DoneWithCustomAction
```

And the function being ran just does
```
unsigned SomeMatchData;
if (match(SomeMatchData))
  apply(SomeMatchData)
```

This approach solves multiple issues in one:
  - MatchData handling is greatly simplified and more efficient, "don't pay for what you don't use"
  - We reduce the size of the match table
  - Calling C++ code has a certain overhead (we need a switch), and this overhead is only paid once now.

Handling of C++ code inside PatFrags is unchanged though, that still emits a `GIM_CheckCxxInsnPredicate`. This is completely fine as they can't use MatchDatas.

### CODE REVIEW

**Comment by jayfoad** [comment]
Awesome! Thanks for doing this so quickly.
```
unsigned SomeMatchData;
if (match(SomeMatchData))
  apply(SomeMatchData)
```
Do the match and apply functions get inlined into this?

**Comment by Pierre-vh** [comment]
> Awesome! Thanks for doing this so quickly.
> 
> ```
> unsigned SomeMatchData;
> if (match(SomeMatchData))
>   apply(SomeMatchData)
> ```
> 
> Do the match and apply functions get inlined into this?

They probably will be inlined if it's possible (simplified example: https://godbolt.org/z/17vY68hx3), but it's all in one big switch so maybe size constraints will prevent everything from being inlined all the time

File: llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h [hunk 1]
```diff
@@ -1335,13 +1335,17 @@ bool GIMatchTableExecutor::executeMatchTable(
...
-                      dbgs() << CurrentIdx << ": GIR_CustomAction(FnID=" << FnID
-                             << ")\n");
+                      dbgs() << CurrentIdx << ": GIR_DoneWithCustomAction(FnID="
+                             << FnID << ")\n");
       assert(FnID > GICXXCustomAction_Invalid && "Expected a valid FnID");
-      runCustomAction(FnID, State, OutMIs);
+      if (runCustomAction(FnID, State, OutMIs)) {
+        propagateFlags();
+        return true;
+      } else if (handleReject() == RejectAndGiveUp)
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h:1347]: No else after return

**Comment by aemerson** [comment]
I'll run the CT tests on this too...

File: llvm/utils/TableGen/GlobalISelCombinerEmitter.cpp [hunk 2]
```diff
@@ -1175,9 +1180,20 @@ bool CombineRuleBuilder::checkSemantics() {
...
   }
 
+  std::optional<bool> IsUsingCXXPatterns;
   for (const auto &Apply : ApplyPats) {
-    assert(Apply.second.get());
-    const auto *IP = dyn_cast<InstructionPattern>(Apply.second.get());
+    Pattern *Pat = Apply.second.get();
+    if (IsUsingCXXPatterns) {
+      if (*IsUsingCXXPatterns != isa<CXXPattern>(Pat)) {
+        PrintError("'apply' patterns cannot mix C++ code with other types of "
```
Comment (arsenm) [inline llvm/utils/TableGen/GlobalISelCombinerEmitter.cpp:1188]: Probably should note this in the documentation

**Comment by aemerson** [comment]
A non-LTO release w/o asserts build:
```
Program                                       compile_time
                                              lhs          rhs    diff
7zip/7zip-benchmark                            14.76        14.75 -0.1%
Bullet/bullet                                  10.98        10.97 -0.1%
tramp3d-v4/tramp3d-v4                           5.68         5.67 -0.1%
kimwitu++/kc                                    6.65         6.64 -0.2%
SPASS/SPASS                                     4.57         4.56 -0.3%
lencod/lencod                                   4.65         4.64 -0.3%
mafft/pairlocalalign                            2.68         2.67 -0.4%
consumer-typeset/consumer-typeset               3.59         3.58 -0.4%
sqlite3/sqlite3                                 2.49         2.48 -0.5%
                           Geomean difference                     -0.3%
```

A huge improvement, fantastic!

File: llvm/utils/TableGen/GlobalISelCombinerEmitter.cpp [hunk 3]
```diff
@@ -1780,6 +1811,44 @@ bool CombineRuleBuilder::emitApplyPatterns(CodeExpansions &CE, RuleMatcher &M) {
...
+         << "  return false;\n}\n";
+    }
+  }
+
+  OS << "// Apply Patterns\n";
+  ListSeparator LS("\n");
+  for (auto &Pat : ApplyPats) {
+    auto *CXXPat = cast<CXXPattern>(Pat.second.get());
+    CodeExpander Expander(CXXPat->getRawCode(), CE, RuleDef.getLoc(),
+                          /*ShowExpansions*/ false);
```
Comment (arsenm) [inline llvm/utils/TableGen/GlobalISelCombinerEmitter.cpp:1842]:
```suggestion
                          /*ShowExpansions=*/ false);
```

File: llvm/utils/TableGen/GlobalISelCombinerEmitter.cpp [hunk 4]
```diff
@@ -1780,6 +1811,44 @@ bool CombineRuleBuilder::emitApplyPatterns(CodeExpansions &CE, RuleMatcher &M) {
...
+
+  for (auto &MD : MatchDatas)
+    OS << MD.Type << " " << MD.getVarName() << ";\n";
+
+  if (!Matchers.empty()) {
+    OS << "// Match Patterns\n";
+    for (auto *M : Matchers) {
+      OS << "if(![&](){";
+      CodeExpander Expander(M->getRawCode(), CE, RuleDef.getLoc(),
+                            /*ShowExpansions*/ false);
```
Comment (arsenm) [inline llvm/utils/TableGen/GlobalISelCombinerEmitter.cpp:1830]:
```suggestion
                            /*ShowExpansions=*/ false);
```

**Review by arsenm — APPROVED** [review]

**Pull request merged**